lib/slack-ruby-bot/commands/base.rb

```diff
@@ -122,6 +131,36 @@ def finalize_routes!
           command command_name_from_class
         end
 
+        def match_attachments(data, route)
+          fields_to_scan = %i[pretext text title]
+          data.attachments.each do |attachment|
+            fields_to_scan.each do |field|
+              next unless attachment[field]
+              match = route.match(attachment[field])
+              next unless match
+              # add access to the matched attachment
+              match.singleton_class.send(
+                :define_method, :attachment, -> { attachment }
+              )
+              match.singleton_class.send(
+                :define_method, :attachment_field, -> { field }
+              )
+              return match
+            end
+          end
+          false
+        end
+
+        def data_missing?(match_method, expression, data)
+          case match_method
+          when :match, :scan
+            return false unless expression.nil?
+          when :attachment
+            return false unless data.attachments.nil? || data.attachments.empty?
+          end
```

Comment:
I prefer extracting logic into methods even if they aren't re-used. Just to increase readability and reduce the cyclomatic complexity. So I decided to re-organize the method to use a case statement.

Reply:
Something doesn't smell right here as we do the case twice, once in `invoke` and once in `data_missing`. Haven't tried to rewrite it, but usually this means both can be combined avoiding having to case twice.

Reply:
I think this is simpler :)

```ruby
case match_method
when ...
  expression.nil?
when :attachment
  data.attachments.nil? || data.attachments.empty
else
  true
end
```

:)

Reply:
We can just put the checks directly in the case in `invoke` but this will increase the cyclomatic complexity.
It would look like
```ruby
when :match
  next unless expression
  # ...
when :scan
  next unless expression
  # ...
when :attachment
  next unless data.attachments && !data.attachments.empty?
  # ...
```

Reply:
I am not a code purist re: cyclomatic complexity :)

Reply:
Okay I'll add the checks to the `invoke` method then :)